### ISSUE TITLE
Edge Missing Transceiver Property Errors

### ISSUE THREAD

**Issue by austinbarnes** [webrtcHacks/adapter#913]
# Please read first!
Please use [discuss-webrtc](https://groups.google.com/forum/#!forum/discuss-webrtc) for general technical discussions and questions.

- [x] I have provided steps to reproduce (e.g. a link to a [jsfiddle](https://jsfiddle.net/))
- [x] I have provided browser name, version and adapter.js version
- [x] This issue only happens when adapter.js is used

**Note: If the checkboxes above are not checked (which you do after the issue is posted), the issue will be closed.**

## Versions affected
Microsoft Edge 42.17134.1.0
Microsoft EdgeHTML 17.17134
Chrome 70.0.3538.110 (Official Build) (64-bit)
adapter.js 7.0.0

## Description
I just recently updated to version 7.0.0 and noticed I was getting some errors while trying to do video calls on Edge. These errors stop the ability for my code to clean up any peers using edge. 

Here are the error messages: 
![adaptererrormessages](https://user-images.githubusercontent.com/5625732/49409203-c02b2c80-f71c-11e8-8abc-64366361f0b8.PNG)

I debugged the errors and found out that there are (see the attached error log) transceivers being added that don't have either the properties iceGatherer or dtlsTransport:
![edgetranscievererror](https://user-images.githubusercontent.com/5625732/49409219-d3d69300-f71c-11e8-8348-4eeee3090e0e.PNG)

The lines that were having errors are: 2566, 2570, 3844, 3882, and 3883
Here are screen shots of some simple if statements I put around them to get the errors not happen:
<img width="1375" alt="screen shot 2018-12-03 at 4 47 03 pm" src="https://user-images.githubusercontent.com/5625732/49408840-56f6e980-f71b-11e8-9dfe-8d2f899d95f5.png">
<img width="1354" alt="screen shot 2018-12-03 at 4 47 30 pm" src="https://user-images.githubusercontent.com/5625732/49408841-56f6e980-f71b-11e8-8382-5eb34af61b1c.png">
<img width="1368" alt="screen shot 2018-12-03 at 4 47 46 pm" src="https://user-images.githubusercontent.com/5625732/49408842-56f6e980-f71b-11e8-95f2-c5a22076c175.png">

## Steps to reproduce
1. Start a call on Edge.
2. Join the call with Chrome on another device.

## Expected results
There are no errors while connecting the video.

## Actual results
There are errors that stop any other functionality like cleaning up peers.

**Comment by fippo** [comment]
Does it still work in 6.4.x? https://github.com/otalk/rtcpeerconnection-shim hasn't seen many (released) changes in the adapter.js 7.0.0 version bump.

This sounds like it is attempting to use writeMediaSection for a (rejected) datachannel mline which isn't supposed to happen. Does this come from createOffer or createAnswer?

**Comment by austinbarnes** [comment]
So the Edge client does call the **createAnswer** and while they are exchanging iceCandidates these errors happen (at least in my console logs). I tried with adapter versions **6.3.0, 6.4.0, 6.4.8** and am still getting these error messages. 

However, I also tried version **6.2.0** and did not get any error messages.

**Comment by fippo** [comment]
I have a fix for the second error (above, thanks!) but can not reproduce the first one.
Do you have either a minimal example or a webrtc-internals dump from your application in a chrome-chrome case?

**Comment by austinbarnes** [comment]
Awesome I am glad you found a fix so quick! So the first error is only happening on Edge. I can get you a chrome-chrome webrtc-internals dump but I don't know how helpful it would be..? Let me work on getting up a simple example as well.

**Comment by austinbarnes** [comment]
Here is the chrome to edge webrtc-internals dump: 
[webrtc_internals_dump.txt](https://github.com/webrtcHacks/adapter/files/2645373/webrtc_internals_dump.txt)

Here is the chrome to chrome webrtc-internals dump: 
[webrtc_internals_dump (1).txt](https://github.com/webrtcHacks/adapter/files/2645449/webrtc_internals_dump.1.txt)

**Comment by fippo** [comment]
hrm... that looks pretty standard and the SDP answer is correctly rejecting the datachannel. Chrome to Edge was done with your local patches, right?
And the flow in Edge is
* pc.setRemoteDescription(offerFromChrome)
* pc.createAnswer()
* pc.setLocalDescription()
and it errors in createAnswer?

I've managed to reproduce that error message in a scenario where
* pc.setRemoteDescription(offerFromChrome)
* pc.createAnswer()
* pc.setLocalDescription()
* pc.createOffer()

the last call to pc.createOffer throws. See https://github.com/otalk/rtcpeerconnection-shim/issues/178

**Comment by austinbarnes** [comment]
The Chrome to Edge was done without my patch. I can do one with my patch if you want?

So the flow on Edge is:
-  receive signaling message with offer
-  pc.setRemoteDescription
-  pc.createAnswer
-  pc.setLocalDescription
-  send the answer to the Chrome client via our signaling piece.

I did just resolve an issue in our code where our onNegotionationNeededEventHandler was firing and trying to create an offer while the peer was already in the process of creating an answer, but I was only getting that error on Chrome and Firefox.

**Comment by fippo** [comment]
hrm... but this looks like you got an answer from Edge and that should not have happened if the error was thrown. :thinking: 

What is the `type`argument to createMediaSection when the error is thrown?

**Comment by austinbarnes** [comment]
So I am a little confused on where to find out the type argument in the code. Is it on the SDP that gets created by the createAnswer?

Also if you fixed the 2nd issue it should resolve the first issue. As soon as I fixed the undefined references to the iceGatherer and dtlsTransport with my simple if statements I didn't get either error messages.

**Comment by fippo** [comment]
Its in
```
function writeMediaSection(transceiver, caps, type, stream, dtlsRole) {
```

interesting if making the event... those events should fire asynchronously so errors in there should not cause subsequent failures.

**Comment by austinbarnes** [comment]
Cool I did some logging of each parameter of that function call without my patch and I got that it got called twice with each type being answer (the dashes separate the calls): 
![writemediasectionlogging](https://user-images.githubusercontent.com/5625732/49524442-fd9cd080-f868-11e8-921b-e199980009a8.PNG)

I am working also on the warning message about the end of candidates message. I saw you had posted on Stackoverflow about adding a null ice candidate in order for that to happen correct? 
article: https://stackoverflow.com/questions/51495599/timeout-for-addremotecandidate-consider-sending-an-end-of-candidates-notificati/51499669#51499669

**Comment by fippo** [comment]
v1.2.15 of https://www.npmjs.com/package/rtcpeerconnection-shim has a patch which should address this issue.

> I am working also on the warning message about the end of candidates message.

You have to send a signal from the peer when it gets to onicecandidate with event.candidate not being defined and then call addIceCandidate(null)

**Comment by fippo** [comment]
should be fixed in v1.2.15 of rtcpeerconnection-shim, please reopen/holler if not
